### PR TITLE
[Core][Actor Cancel] Recursive Cancelation.

### LINKED ISSUE #38711: [CI] `osx://python/ray/tests:test_actor_cancel` is failing/flaky on master.
- dea8ad7cb7dd98eb5a0d75a053847ce282f84230 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5601#018a1a41-e33c-48ab-ab8f-bfce6d0e87a5)
- a520d863a45a02d0bfa0f95f6392636969e9b783 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5600#018a1a24-c7d7-47ce-a86a-637d16570810)
- 56d66677b175ffab366ff37074285af1356e34c0 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5598#018a1a17-8a59-488e-81b2-0dbe80de23a0)
- e2c6fc5e36aeaa2a8ede83500a01b4992702033e FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5597#018a1a12-d2e2-40d5-82c5-903dbbeae3f0)
- 34578a69216fbf25855677841f9b25a5910409c7 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5596#018a19ea-9c0a-406d-bb67-a1dfe373555f)
- 299ae25a048f05ab74afb4b69a80f51014b42029 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5594#018a19c6-a5e4-48f1-951c-fea99b121258)
- 310da65382d2e3ddd037ddfc605821be31b32eee FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5593#018a19c2-4ced-4f85-b435-6fd7254f95a6)
- 4b37161d30b3c538b55f1d661248fe43078d19f8 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5592#018a19ba-55f6-47d2-8f5a-875be82fb72b)
- 2d470954f2855485da5f34eddbb8bcfae441c3b4 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5591#018a19aa-d730-4003-9251-c29fad167236)
- b939a56991dec29874f293a9b57aa2124d2454e1 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5590#018a1973-b7aa-49bc-a07c-d25c8666f639)
- a284319fd2a1bd5721b5e35d4bc02436416ea1c9 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5589#018a196a-15f3-42a3-9be8-b3439c988c02)
- f6bcaf878620b433ebbb475a74699d6f201f92b6 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5588#018a1966-157b-4a6b-ab2d-def45722529b)
- 1efd2e280d7a424d18d5afb8ca822fdfed966980 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5587#018a193d-ee21-45ae-b735-df1718f750bd)
- f8e5ac298de95429f73faec3df2274d4053eed68 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5586#018a1934-8df3-4cf7-bf8d-8104e5379ebf)
- 87181ce3d6b1e538a9c672d71b8871f8e2fe2adb FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5585#018a1911-41b0-4645-8d13-d0534a30ba3e)
- ed443252a67bedcd862d53592eded974770780a7 FAILED [Buildkite :mac: :apple: Medium A-J](https://buildkite.com/ray-project/oss-ci-build-branch/builds/5584#018a1909-05c1-42d6-aee2-5a8f832134f1)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-osx://python/ray/tests:test_actor_cancel-END
....

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR supports recursive actor task cancelation. 

- Introduce the contextvar to allow to obtain correct task_id
- Correctly return task_id for get_runtime_context().get_task_id() inside an async task.
- Correctly set parent_task_id for asyncio tasks when submitting a task. This will help us correctly figure out children tasks to cancel.
- Support recursive cancelation
- Testing.

Note that this fixes the asyncio context issue temporarily. We should make a correct fix by ray 2.8. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/38711

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 1]
```diff
@@ -292,7 +292,12 @@ void CoreWorkerDirectTaskReceiver::RunNormalTasksFromQueue() {
 bool CoreWorkerDirectTaskReceiver::CancelQueuedActorTask(const WorkerID &caller_worker_id,
                                                          const TaskID &task_id) {
   auto it = actor_scheduling_queues_.find(caller_worker_id);
-  return it->second->CancelTaskIfFound(task_id);
+  if (it != actor_scheduling_queues_.end()) {
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_transport.cc:295]: This is the edge case I found while running a test.

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -3444,6 +3454,11 @@ cdef class CoreWorker:
             CSchedulingStrategy c_scheduling_strategy
             c_vector[CObjectID] incremented_put_arg_ids
             c_string serialized_retry_exception_allowlist
+            CTaskID current_c_task_id
+            TaskID task_id_in_async_context = async_task_id.get()
+            # This task id is incorrect if async task is used.
```
Comment (jjyao) [inline python/ray/_raylet.pyx:3459]: Can we always use contextvar for all types of actors not just async actors?

File: python/ray/_raylet.pyx [hunk 3]
```diff
@@ -4258,6 +4303,25 @@ cdef class CoreWorker:
         with nogil:
             CCoreWorkerProcess.GetCoreWorker().YieldCurrentFiber(fiber_event)
 
+    def get_pending_children_task_ids(self, parent_task_id: TaskID):
```
Comment (jjyao) [inline python/ray/_raylet.pyx:4306]: Is this just for testing?

**Review by jjyao — COMMENTED** [review]
lg

File: python/ray/_raylet.pyx (hunk 3 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:4306]: Yes! I will add comments.

File: python/ray/_raylet.pyx (hunk 2 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:3459]: I think it is possible, but if it s not a hard blocker, I prefer to properly fix it when we fix the whole thing in 2.8.

The reason is this requires us to properly "unset" task_id in various scenarios (e.g., task raises an exception). This is doable, but it requires more tests and stuff which increases the scope of the PR unnecessarily.

**Review by jjyao — APPROVED** [review]

**Review by rickyyx — APPROVED** [review]

**Comment by rkooo567** [comment]
Failed tests seem unrelated

**Pull request merged**